source/extensions/filters/http/sxg/filter.h

```diff
@@ -0,0 +1,207 @@
+#pragma once
+
+#include "envoy/server/filter_config.h"
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "envoy/extensions/filters/http/sxg/v3alpha/sxg.pb.h"
+
+#include "source/common/config/datasource.h"
+
+#include "libsxg.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace SXG {
+
+#define ALL_SXG_STATS(COUNTER)                                                                     \
+  COUNTER(total_client_can_accept_sxg)                                                             \
+  COUNTER(total_should_sign)                                                                       \
+  COUNTER(total_exceeded_max_payload_size)                                                         \
+  COUNTER(total_signed_attempts)                                                                   \
+  COUNTER(total_signed_succeeded)                                                                  \
+  COUNTER(total_signed_failed)
+
+struct SignedExchangeStats {
+  ALL_SXG_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+// Helper class used to fetch secrets (usually from SDS).
+class SecretReader {
+public:
+  virtual ~SecretReader() = default;
+  virtual const std::string& certificate() const PURE;
+  virtual const std::string& privateKey() const PURE;
+};
+
+class SDSSecretReader : public SecretReader {
+public:
+  SDSSecretReader(Secret::GenericSecretConfigProviderSharedPtr certificate_provider,
+                  Secret::GenericSecretConfigProviderSharedPtr private_key_provider, Api::Api& api)
+      : update_callback_client_(readAndWatchSecret(certificate_, certificate_provider, api)),
+        update_callback_token_(readAndWatchSecret(private_key_, private_key_provider, api)) {}
+
+  const std::string& certificate() const override { return certificate_; }
+
+  const std::string& privateKey() const override { return private_key_; }
```

Comment:
nit: group these together and when overriding it's usually nice to have a comment referencing the parent for those methods, e.g.:

```
// SecretReader
const std::string& certificate() const override { return certificate_; }
const std::string& privateKey() const override { return private_key_; }
```